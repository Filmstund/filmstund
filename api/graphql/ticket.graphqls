extend type Showing {
    myTickets: [Ticket!]!
    ticketRange: TicketRange
}

extend type Mutation {
    processTicketUrls(showingID: UUID!, ticketUrls: [String!]): Showing! @auth(requires: SHOWING_ADMIN)
}

# A ticket assigned to a user that allows access to Filmstaden
type Ticket {
    id: String!
    showingID: UUID!
    assignedToUser: UUID!
    profileID: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seat: Seat!
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]!
}

type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]!
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]!
    seatings: [SeatRange!]!
    # The total number of tickets for this showing
    totalCount: Int!
}
