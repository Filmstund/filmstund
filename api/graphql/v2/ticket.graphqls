extend type Showing {
    myTickets: [Ticket!]!
    ticketRange: TicketRange
}

# A ticket assigned to a user that allows access to Filmstaden
type Ticket {
    id: String!
    showingID: UUID!
    assignedToUser: UserID!
    profileID: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seat: Seat!
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]!
}

type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]!
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]!
    seatings: [SeatRange!]!
    # The total number of tickets for this showing
    totalCount: Int!
}
