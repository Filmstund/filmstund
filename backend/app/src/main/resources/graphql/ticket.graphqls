# A ticket assigned to a user that allowes access to Filmstaden
type Ticket {
    id: String!
    #showingId: UUID! FIXME use dto where this exists instead
    assignedToUser: User!
    profileId: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seatNumber: Int!,
    seatRow: Int!,
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    attributes: [String!]
}

# The row and seat number for a ticket
type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]
    seatings: [SeatRange!]
    # The total number of tickets for this showing
    totalCount: Int!
}

extend type Showing {
    myTickets: [Ticket!]
    ticketRange: TicketRange
}