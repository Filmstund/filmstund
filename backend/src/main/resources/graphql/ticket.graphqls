# A ticket assigned to a user that allowes access to SF
type Ticket {
    id: String!
    showingId: UUID!
    assignedToUser: User!
    profileId: String
    barcode: String!
    customerType: String!
    customerTypeDefinition: String!
    cinema: String!
    cinemaCity: String
    screen: String!
    seat: Seat!,
    date: LocalDate!,
    time: String!,
    movieName: String!
    # 15 år, 11 år etc.
    movieRating: String!
    # "textad", "en" etc
    showAttributes: [String!]
}

# The row and seat number for a ticket
type Seat {
    row: Int!
    number: Int!
}

type SeatRange {
    row: Int!
    # All numbers for this particular row
    numbers: [Int!]
}

# In what range is the seating assigned for each row
type TicketRange {
    rows: [Int!]
    seatings: [SeatRange!]
    # The total number of tickets for this showing
    totalCount: Int!
}

extend type Showing {
    myTickets: [Ticket!]
    ticketRange: TicketRange
}