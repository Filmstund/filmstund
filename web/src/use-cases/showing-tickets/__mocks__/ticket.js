export const createMockSeat = (row, number, x, y) => ({
  row,
  number,
  coordinates: {
    x,
    y
  },
  dimensions: {
    width: 10,
    height: 10
  }
});

export const createMockTicket = (id, row, number) => ({
  id,
  barcode: "this-is-barcode",
  customerType: "mock-customer-type",
  customerTypeDefinition: "mock-customer-type-definition",
  cinema: "mock-cinema",
  screen: "mock-screen",
  profileId: "mock-profile-id",
  seat: {
    row,
    number
  },
  date: "2018-10-10",
  time: "20:00",
  movieName: "mock-movie-name",
  movieRating: "R",
  showAttributes: ["3D"]
});
