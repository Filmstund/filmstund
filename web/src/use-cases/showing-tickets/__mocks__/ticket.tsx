export const createMockSeat = (
  row: number,
  number: number,
  x: number,
  y: number
) => ({
  row,
  number,
  coordinates: {
    x,
    y,
  },
  dimensions: {
    width: 10,
    height: 10,
  },
});

export const createMockTicket = (id: string, row: number, number: number) => ({
  id,
  barcode: "this-is-barcode",
  customerType: "mock-customer-type",
  customerTypeDefinition: "mock-customer-type-definition",
  cinema: "mock-cinema",
  screen: "mock-screen",
  profileId: "mock-profile-id",
  seat: {
    row,
    number,
  },
  date: "2018-10-10",
  time: "20:00",
  movieName: "mock-movie-name",
  movieRating: "R",
  showAttributes: ["3D"],
});
