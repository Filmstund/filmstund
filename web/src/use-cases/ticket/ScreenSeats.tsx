import styled from "@emotion/styled";
import { flatMap, groupBy, map } from "lodash";
import React from "react";
import { TicketFragment } from "../../__generated__/types";

const Seat = styled.div<{ selected: boolean }>(({ selected }) => ({
  backgroundColor: selected ? "#d0021b" : "#9b9b9b",
  position: "absolute",
  borderRadius: 2,
}));

const Screen = styled.div<{ height: number; width: number }>`
  margin: 20px 0;
  position: relative;
  height: ${(props) => props.height}px;
  width: ${(props) => props.width}px;
`;

interface Props {
  ticketRange: TicketFragment["ticketRange"];
  seatMap: TicketFragment["filmstadenSeatMap"];
}

export const ScreenSeats: React.FC<Props> = ({ ticketRange, seatMap }) => {
  ticketRange = ticketRange!;
  if (seatMap.length === 0) {
    return null;
  }

  const rows = groupBy(seatMap, (s) => s.row);
  const widestRowLength = Math.max(...map(rows, (row) => row.length));

  const width = Math.min(window.innerWidth - 40, window.innerHeight, 650);

  const optimalSeatWidth = 17;
  const seatWidth = Math.min(optimalSeatWidth, width / (widestRowLength * 1.2));
  const seatHeight = Math.floor((13 * seatWidth) / optimalSeatWidth);

  const minY = Math.min(...seatMap.map((s) => s.coordinates.y));
  const maxY = Math.max(...seatMap.map((s) => s.coordinates.y));
  const minX = Math.min(...seatMap.map((s) => s.coordinates.x));
  const maxX = Math.max(...seatMap.map((s) => s.coordinates.x));

  const scale = width / (maxX - minX);

  const numbers = flatMap(
    ticketRange.seatings,
    (seating) => seating.numbers || []
  );

  return (
    <Screen
      height={(maxY - minY) * scale + seatHeight}
      width={(maxX - minX) * scale + seatWidth}
    >
      {seatMap.map((data) => (
        <Seat
          style={{
            left: Math.floor((data.coordinates.x - minX) * scale),
            top: Math.floor((data.coordinates.y - minY) * scale),
            height: seatHeight,
            width: seatWidth,
          }}
          key={data.number}
          selected={numbers.includes(data.number)}
        />
      ))}
    </Screen>
  );
};
