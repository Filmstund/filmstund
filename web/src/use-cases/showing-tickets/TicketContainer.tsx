import React, { useState } from "react";
import MainButton from "../common/ui/MainButton";
import TicketURLInput from "../../use-cases/common/ui/TicketURLInput";
import { navigateToShowing } from "../common/navigators";

import { FieldWithoutMaxWidth } from "../common/ui/Field";
import { SmallHeader } from "../common/ui/Header";
import { ScreenSeats } from "../ticket/ScreenSeats";
import {
  useAddTicketsMutation,
  useTicketQuery,
} from "../../__generated__/types";
import { SeatRange } from "./SeatRange";
import { Ticket } from "./Ticket";
import { useNavigate } from "react-router-dom";
import { useToaster } from "../../common/toast/ToastContext";

interface Props {
  webID: string;
}

export const TicketContainer: React.FC<Props> = ({ webID }) => {
  const [{ data }] = useTicketQuery({ variables: { webID } });
  const { me, showing } = data!;

  if (!showing) {
    throw new Error("Showing is missing");
  }

  const navigate = useNavigate();

  const [cinemaTicketUrls, setCinemaTicketUrls] = useState<string[]>([]);
  const [, addTickets] = useAddTicketsMutation();
  const toast = useToaster();

  const handleGoBackToShowing = () => {
    navigateToShowing(navigate, showing);
  };

  const handleSubmitCinemaTicketUrls = () => {
    const nonEmptyUrls = cinemaTicketUrls.filter(
      (line) => line.trim().length !== 0
    );

    addTickets({ showingId: showing.id, tickets: nonEmptyUrls }).then(
      ({ data, error }) => {
        setCinemaTicketUrls([]);
        if (data) {
          toast({ text: "Uppdaterades!", variant: "success" });
        } else if (error) {
          toast({ text: error.message, variant: "danger" });
        }
      }
    );
  };

  const { myTickets, ticketRange, filmstadenSeatMap } = showing;

  const tickets = myTickets || [];

  return (
    <>
      <MainButton onClick={handleGoBackToShowing}>
        Tillbaka till visning
      </MainButton>
      <SmallHeader>Våra platser:</SmallHeader>
      <SeatRange ticketRange={ticketRange} />
      <div style={{ display: "flex", justifyContent: "center" }}>
        {ticketRange && (
          <ScreenSeats ticketRange={ticketRange} seatMap={filmstadenSeatMap} />
        )}
      </div>
      {tickets.map((ticket) => (
        <Ticket key={ticket.id} ticket={ticket} />
      ))}
      {showing.admin.id === me.id && (
        <FieldWithoutMaxWidth text="Lägg till Filmstaden-bokningslänkar">
          <TicketURLInput
            cinemaTicketUrls={cinemaTicketUrls}
            onChange={setCinemaTicketUrls}
          />
          <MainButton onClick={handleSubmitCinemaTicketUrls}>Skicka</MainButton>
        </FieldWithoutMaxWidth>
      )}
    </>
  );
};
