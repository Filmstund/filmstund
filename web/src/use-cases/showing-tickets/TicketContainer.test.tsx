import React from "react";
import { cleanup, render } from "@testing-library/react";

import { TicketContainer } from "./TicketContainer";
import { createMockSeat, createMockTicket } from "./__mocks__/ticket";
import { MemoryRouter as Router } from "react-router-dom";
import {
  ApolloClient,
  ApolloLink,
  ApolloProvider,
  InMemoryCache,
} from "@apollo/client";
import {
  TicketQuery_me,
  TicketQuery_showing,
} from "./__generated__/TicketQuery";

const mockUserData: TicketQuery_me = {
  id: "this-is-mock-user-id",
} as any;

const minimalMockData: TicketQuery_showing = {
  id: "this-is-mock-showing-id",
  admin: mockUserData,
  ticketRange: {
    rows: [],
    seatings: [],
  },
  filmstadenSeatMap: [],
  myTickets: [createMockTicket("1", 2, 3)],
} as any;

const mockData: TicketQuery_showing = {
  id: "this-is-mock-showing-id",
  admin: mockUserData,
  ticketRange: {
    rows: [2, 3],
    seatings: [
      {
        row: 2,
        numbers: [3, 4],
      },
      {
        row: 3,
        numbers: [5, 6],
      },
    ],
  },
  filmstadenSeatMap: [
    createMockSeat(2, 3, 5, 0),
    createMockSeat(2, 4, 10, 0),
    createMockSeat(3, 5, 5, 5),
    createMockSeat(3, 6, 10, 5),
  ],
  myTickets: [
    createMockTicket("1", 2, 3),
    createMockTicket("2", 2, 4),
    createMockTicket("3", 3, 5),
    createMockTicket("4", 3, 6),
  ],
} as any;

afterEach(cleanup);

const createEmptyApolloClient = () =>
  new ApolloClient({ link: ApolloLink.empty(), cache: new InMemoryCache() });

describe("<TicketContainer />", () => {
  it("renders without crashing", () => {
    const { container } = render(
      <ApolloProvider client={createEmptyApolloClient()}>
        <Router>
          <TicketContainer me={mockUserData} showing={minimalMockData} />
        </Router>
      </ApolloProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the correct amount of tickets", () => {
    const { queryAllByText } = render(
      <ApolloProvider client={createEmptyApolloClient()}>
        <Router>
          <TicketContainer me={mockUserData} showing={mockData} />
        </Router>
      </ApolloProvider>
    );
    expect(queryAllByText("mock-cinema")).toHaveLength(4);
  });
});
