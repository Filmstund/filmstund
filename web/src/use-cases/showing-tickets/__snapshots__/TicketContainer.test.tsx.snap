// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketContainer /> renders without crashing 1`] = `
<div>
  <button
    class="css-gcz4q9"
  >
    Tillbaka till visning
  </button>
  <h3
    class="css-1llnadu"
  >
    Våra platser:
  </h3>
  <div
    style="display: flex; justify-content: center;"
  />
  <div
    class="css-18sytmz"
  >
    <div
      class="css-o2j9ze"
      style="margin-bottom: 1rem;"
    >
      <img
        alt="FS logo"
        src=""
        style="height: 3rem; width: auto; margin-right: 1rem;"
      />
      <div
        class="css-deizzc"
      >
        <div
          class="css-lp25y4"
          style="font-size: 1rem;"
        >
          mock-cinema
        </div>
        <div
          class="css-e2s1aj"
        >
          Filmstaden
        </div>
      </div>
    </div>
    <div
      class="css-deizzc"
      style="margin-bottom: 1rem;"
    >
      <div
        class="css-lp25y4"
        style="margin-bottom: 0.5rem;"
      >
        mock-movie-name
      </div>
      <div
        class="css-f24dyb"
      >
        <div
          class="css-o2j9ze"
        >
          <span
            class="css-ya7yf4"
          >
            3D
          </span>
        </div>
        R
      </div>
    </div>
    <div
      class="css-o2j9ze"
      style="margin-bottom: 1rem;"
    >
      <div
        class="css-deizzc"
        style="padding-right: 1rem;"
      >
        <div
          class="css-e2s1aj"
        >
          Datum
        </div>
        <div
          class="css-lp25y4"
        >
          2018-10-10
        </div>
      </div>
      <div
        class="css-deizzc"
        style="padding-right: 1rem;"
      >
        <div
          class="css-e2s1aj"
        >
          Tid
        </div>
        <div
          class="css-lp25y4"
        >
          20:00
        </div>
      </div>
    </div>
    <div
      class="css-1k01z69"
      style="margin-bottom: 1rem; border: .0625rem solid #000;"
    >
      <div
        class="css-deizzc"
        style="width: 50%; padding-right: 1rem;"
      >
        <div
          class="css-e2s1aj"
        >
          Salong
        </div>
        <div
          class="css-lp25y4"
        >
          mock-screen
        </div>
      </div>
      <div
        class="css-deizzc"
        style="padding-right: 1rem;"
      >
        <div
          class="css-e2s1aj"
        >
          Rad
        </div>
        <div
          class="css-lp25y4"
        >
          2
        </div>
      </div>
      <div
        class="css-deizzc"
        style="padding-right: 1rem;"
      >
        <div
          class="css-e2s1aj"
        >
          Stolsnr
        </div>
        <div
          class="css-lp25y4"
        >
          3
        </div>
      </div>
    </div>
    <div
      class="css-o2j9ze"
      style="margin-bottom: 1rem; font-weight: 500;"
    >
      mock-customer-type
    </div>
    <div
      class="css-o2j9ze"
      style="margin-bottom: 1rem;"
    >
      <div
        class="css-deizzc"
        style="align-items: left;"
      >
        <img
          alt="1"
          class="css-6n2ytg"
          src="this-is-barcode"
        />
        <div>
          1
           
          mock-profile-id
        </div>
      </div>
    </div>
  </div>
  <div
    class="css-1dhwn8t"
  >
    <label
      class="css-1vtr8wv"
    >
      Lägg till Filmstaden-bokningslänkar
    </label>
    <br />
    <textarea
      class="css-1tvymxn"
      rows="5"
    />
    <button
      class="css-gcz4q9"
    >
      Skicka
    </button>
  </div>
</div>
`;
